### PR TITLE
blog: add post about my porting of the feather-nrf52840

### PR DESCRIPTION
**Pull request by miri64**
As discussed during the [VMA 2021.02](https://forum.riot-os.org/t/virtual-maintainer-assembly-2021-02/3111/9), I am offering my blog article from February 2020 about porting the feather-nrf52840 to RIOT.

### CODE REVIEW

**Comment by miri64** [comment]
In addition to that post (or alternatively to it) I also have [this post](https://blog.martine-lenders.eu/xtimer-en.html) that might be of interest.

**Comment by miri64** [comment]
BTW, I noticed that syntax highlighting for code segments currently does not work. Is there a way to enable it?

File: _posts/2021-02-26-feather-nrf52840.md [hunk 1]
```diff
@@ -0,0 +1,425 @@
+---
+layout: blog
+title: How to Port a Board for RIOT
+authors: Martine S. Lenders
```
Comment (miri64) [inline _posts/2021-02-26-feather-nrf52840.md:4]: BTW, this isn't shown (and I am not sure I am using the right meta info here... just copied what I had in Pelican). IMHO if we want blog posts by external authors, some attribution would be nice ;-).

**Comment by waehlisch** [comment]
> BTW, I noticed that syntax highlighting for code segments currently does not work. Is there a way to enable it?

yes, https://jekyllrb.com/docs/liquid/tags/#code-snippet-highlighting

File: _posts/2021-02-26-feather-nrf52840.md (hunk 1 above)
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:4]: can you also add:
```
githubuid: miri64
```

**Comment by miri64** [comment]
> > BTW, I noticed that syntax highlighting for code segments currently does not work. Is there a way to enable it?
> 
> yes, https://jekyllrb.com/docs/liquid/tags/#code-snippet-highlighting

Did that but does not seem to change anything.

**Comment by waehlisch** [comment]
> > > BTW, I noticed that syntax highlighting for code segments currently does not work. Is there a way to enable it?
> > 
> > 
> > yes, https://jekyllrb.com/docs/liquid/tags/#code-snippet-highlighting
> 
> Did that but does not seem to change anything.

did you check that the CSS definitions are available?

File: _posts/2021-02-26-feather-nrf52840.md (hunk 1 above)
Comment (leandrolanzieri) [reply _posts/2021-02-26-feather-nrf52840.md:4]: See #21

**Comment by miri64** [comment]
> did you check that the CSS definitions are available?

No, did not know that needed to be provided manually. Did that now and provided a RIOT-ified version of the [monokai](https://jwarby.github.io/jekyll-pygments-themes/languages/ruby.html) theme (which was already pretty RIOT-ty which is why I picked it)

**Comment by waehlisch** [comment]
> > did you check that the CSS definitions are available?
> 
> No, did not know that needed to be provided manually.

i suppose that is why the documentation says: "Copy the CSS file (native.css for example) into your css directory and import the syntax highlighter styles into your main.css:"

**Comment by miri64** [comment]
> i suppose that is why the documentation says: "Copy the CSS file (native.css for example) into your css directory and import the syntax highlighter styles into your main.css:"

I somehow did not get this far in the article ;-).

**Comment by miri64** [comment]
Added some changes, that @leandrolanzieri suggested offline.

**Comment by miri64** [comment]
Rebased, squashed and adopted for #21.

File: _posts/2021-02-26-feather-nrf52840.md [hunk 2]
```diff
@@ -0,0 +1,462 @@
+---
+layout: blog
+title: How to Port a Board for RIOT
+authors:
+- name: Martine S. Lenders
+  email: mlenders@riot-os.org
+  affiliation: Freie Universität Berlin
+  githubid: miri64
+  twitterid: miri_64
+  stackoverflowid: 11921757
```
Comment (miri64) [inline _posts/2021-02-26-feather-nrf52840.md:10]: There is a bug that we did not spot in #21 for this link. See https://github.com/RIOT-OS/riot-os.org/pull/21#pullrequestreview-600545886. Should I just fix it here, or do we do this in a separate PR?

Comment (leandrolanzieri) [reply _posts/2021-02-26-feather-nrf52840.md:10]: Feel free to piggyback that here

File: _posts/2021-02-26-feather-nrf52840.md [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+the directory is indeed describing a board. This file is called
+[`boards/feather-nrf52840/Makefile`][Makefile]:
+
+{% highlight Makefile %}
+MODULE = board
+
+include $(RIOTBASE)/Makefile.base
+{% endhighlight %}
+
+The first line tells the build system, that this is a module called `board` (all
```
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:91]: usually, there is no comma before "that": system that

File: _posts/2021-02-26-feather-nrf52840.md [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+boards are called like this) and the third includes all the stuff the build
+system developers of RIOT already provided for a module to be built.
+
+However, the build system needs more information. What features (i.e. interfaces
+and devices) does the board support? This is done with the file
```
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:96]: This is defined in the

File: _posts/2021-02-26-feather-nrf52840.md [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+FEATURES_PROVIDED += periph_uart
+FEATURES_PROVIDED += periph_usbdev
+
+# Various other features (if any)
+FEATURES_PROVIDED += radio_nrf802154
+
+include $(RIOTBOARD)/common/nrf52/Makefile.features
+{% endhighlight %}
+
+The first line tells the build system, the CPU (or more accurate microcontoller)
```
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:114]: tells the build system which CPU ...

File: _posts/2021-02-26-feather-nrf52840.md [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+the board is using in a naming scheme that is understand by the build system. We
```
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:115]: is using, applying a naming scheme that the build system understands.

File: _posts/2021-02-26-feather-nrf52840.md [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+basically tell it, "this board is using an nRF52840 microcontroller". Lines 4-9
```
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:116]: Basically, we configure "...".

File: _posts/2021-02-26-feather-nrf52840.md [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+communication interfaces such as I²C, SPI, UART, and USB, but also that the
+radio of this board can speak the [IEEE 802.15.4] protocol—imagine something
+like low-power WiFi—via its nRF microcontroller.
+The last line declares to include further features provided by the nRF52 family
+of microcontrollers in general.
+
+Lastly, I tell the build system with the file
+[`boards/feather-nrf52840/Makefile.include`][Makefile.include], what else needs
+to be included for this board to be programmable, among them the most important
+thing: the programmer itself. In case of the Feather nRF52840 this is the Segger
```
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:127]: i don't understand the last part of the sentence: the programmer itself?

File: _posts/2021-02-26-feather-nrf52840.md [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+ifeq (,$(filter stdio_%,$(DISABLE_MODULE) $(USEMODULE)))
+  RIOT_TERMINAL ?= jlink
+endif
+
+include $(RIOTMAKE)/tools/serial.inc.mk
+
+include $(RIOTBOARD)/common/nrf52/Makefile.include
+{% endhighlight %}
+
+Now we finally leave the build system and tell the microcontroller where to put
```
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:143]: Now, ... 

"tell the microcontroller where to put all the peripherical interfaces to the outside world" -- needs rephrasing.

File: _posts/2021-02-26-feather-nrf52840.md [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+#endif
+
+#endif /* PERIPH_CONF_H */
+{% endhighlight %}
+
+We also need to tell the sensor/actuator abstraction layer of RIOT, [SAUL],
+where to find all the buttons and LEDs provided on the board, so it can react to
+button push events and control the LEDs. This we do in
+[`boards/feather-nrf52840/include/gpio_params.h`][gpio_params.h]. There are
+two LEDs and one switch and I again just followed the [schematics]:
```
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:246]: switch, and I ...

File: _posts/2021-02-26-feather-nrf52840.md [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* GPIO_PARAMS_H */
+/** @} */
+{% endhighlight %}
+
+Lastly, the board needs to define some low-level debugging LED and button [GPIO]
```
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:310]: our board definition needs to include

File: _posts/2021-02-26-feather-nrf52840.md [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+    gpio_clear(LED1_PIN);
+
+    /* initialize the CPU */
+    cpu_init();
+}
+
+/** @} */
+{% endhighlight %}
+
+After that I tested if everything worked as configured, [provided some
```
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:389]: After that the most important steps are needed ;): I tested ...

File: _posts/2021-02-26-feather-nrf52840.md [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+documentation how to do that][feather doc], and then called it done.
+
+# Conclusion
+Given that you already worked with C and `make` and that the microcontroller for
+the board you want to port is already provided, porting a board is quite
```
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:394]: you might want to link to the Boards page

File: _posts/2021-02-26-feather-nrf52840.md [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
+straight-forward. Basically, all that is needed are 7 very concise files. All
```
Comment (waehlisch) [inline _posts/2021-02-26-feather-nrf52840.md:395]: seven very cocise files, <name them>.

**Review by waehlisch — CHANGES_REQUESTED** [review]
i gave a *very* brief read and suggested some changes in terms of wording. the proofread is not complete

File: _posts/2021-02-26-feather-nrf52840.md [hunk 15]
```diff
@@ -0,0 +1,457 @@
...
+image: feather-nrf52840.jpg
+---
+
+*This was originally published on [Martine's blog in February
+2020][original-post].*
+
+# Introduction
+
+It is already a few weeks back, but this year marked a debut for me:
+Though I am a long-term contributor to RIOT—my oldest
```
[the remark this answers is not in the document]
Comment (miri64) [reply _posts/2021-02-26-feather-nrf52840.md:25]: This is "A---B", due to monospace the mdash just looks very squashed.

File: _posts/2021-02-26-feather-nrf52840.md [hunk 16]
```diff
@@ -0,0 +1,457 @@
+---
+layout: blog
+title: How to Port a Board for RIOT
+authors: Martine S. Lenders
+githubuid: miri64
+date: 2021-02-26
+tags: riot, board, microcontroller, adafruit, feather, nrf52
+description:
+    Our contributor Martine Lenders documents her experience of porting her very
```
[the remark this answers is not in the document]
Comment (miri64) [reply _posts/2021-02-26-feather-nrf52840.md:14]: I'll keep the contributor part. In this context I think it is unimportant that I am a maintainer.

[the remark this answers is not in the document]
Comment (waehlisch) [reply _posts/2021-02-26-feather-nrf52840.md:14]: it is not. but anyhow, your choice.

File: _posts/2021-02-26-feather-nrf52840.md [hunk 17]
```diff
@@ -0,0 +1,457 @@
...
+system for hardware platforms that do not typical PC operating systems fit like
+Linux, Mac OSX, or Windows or even phone operating systems like Android or iOS.
+When talking of IoT devices, the RIOT community usually thinks in terms of
+[kilobytes of memory][RFC 7228]. Not gigabytes, not megabytes, _kilobytes_! As
+in, a few 1000 _bytes_. The premise behind this thinking is, that IoT devices
+can not follow [Moore's Law]—which states that every two years the number of
+transistors in a given piece of computer hardware approximately doubles. In
+fact, the efforts to save cost and energy for IoT hardware are directly opposed
+to Moore's Law: every additional transistor, be it for calculation or storage,
+costs more money and leads to more energy usage. Given, that for the IoT a
```
[the remark this answers is not in the document]
Comment (miri64) [reply _posts/2021-02-26-feather-nrf52840.md:39]: Is that the case? My interpretation of Moore's Law was always, that this is exactly not the case. Otherwise, one could argue that Moore's Law would still hold true if would have vacuum tubes in skyscraper sized computers.

File: _posts/2021-02-26-feather-nrf52840.md (hunk 8 above)
Comment (miri64) [reply _posts/2021-02-26-feather-nrf52840.md:127]: Yes, `Makefile.include` informs the build system which programmer to use for `make flash`.

File: _posts/2021-02-26-feather-nrf52840.md (hunk 13 above)
Comment (miri64) [reply _posts/2021-02-26-feather-nrf52840.md:394]: Why? What context does that provide? Or do you mean the CPU page for "Given […] that the microcontroller for the board you want to port is already provided, […]"

File: _posts/2021-02-26-feather-nrf52840.md (hunk 2 above)
Comment (miri64) [reply _posts/2021-02-26-feather-nrf52840.md:10]: Fix provided in #32

File: _posts/2021-02-26-feather-nrf52840.md [hunk 18]
```diff
@@ -0,0 +1,462 @@
...
+that area is mostly the interaction with the network device. Thus, there was
+just never the need for me to port a board for RIOT. So why did I do it now?
+Mainly, because I wanted to try and learn and because I had this piece of
+hardware lying around, for which I knew, that the port would be relatively easy,
+considering my skill set.
+
+The board I ported is the [Adafruit Feather nRF52840]. I bought this board
+earlier this year mainly to save some shipping costs, but also because I like
+the form factor of Adafruit's Feather line and wanted to see more support for it
+in RIOT.  It then became incidentally part of the most current RIOT release
```
[the remark this answers is not in the document]
Comment (miri64) [reply _posts/2021-02-26-feather-nrf52840.md]: Put the fact, that the board was merged into the succeeding release into its own paragraph and clarified the context.

File: _posts/2021-02-26-feather-nrf52840.md [hunk 19]
```diff
@@ -0,0 +1,462 @@
...
+
+I am intentionally leaving the process of [creating a pull request] out
+as I want to focus this article mainly on how to provide a board port, not how
+to get it upstream.
+
+# Creating the board port
+As I said, a board in RIOT is mostly just configuration, both for the build
+system and the actual code. But first I needed a place to put this
+configuration.
+For simplicity I cut out the “Adafruit” out of the board's identifying name—and
```
[the remark this answers is not in the document]
Comment (miri64) [reply _posts/2021-02-26-feather-nrf52840.md]: Re-ordered and split the sentence. It now reads "I cut out the “Adafruit” out of the board's identifying name. Both for simplicity and"

File: _posts/2021-02-26-feather-nrf52840.md (hunk 3 above)
Comment (miri64) [reply _posts/2021-02-26-feather-nrf52840.md]: I know ;-) Appearently, I wasn't that used to _that_ a year ago.

File: _posts/2021-02-26-feather-nrf52840.md (hunk 13 above)
Comment (waehlisch) [reply _posts/2021-02-26-feather-nrf52840.md:401]: CPUs page (and maybe Boards to check whether the Board is already supported).

Comment (miri64) [reply _posts/2021-02-26-feather-nrf52840.md:401]: >  (and maybe Boards to check whether the Board is already supported).

Doesn't really fit in this context here. This should be the implied step 1 ;-) (which I won't editorialize in, just to mention the boards page now).

Comment (miri64) [reply _posts/2021-02-26-feather-nrf52840.md]: Linked the CPUs page.

**Comment by waehlisch** [comment]
@miri64 can you rebase?

**Comment by miri64** [comment]
Done.

**Review by waehlisch — APPROVED** [review]

**Comment by miri64** [comment]
Thanks for the review and merging.
